### ISSUE TITLE
gradient of Dirichlet.log_prob gives nan

### ISSUE THREAD

**Issue by torabin** [pytorch/pytorch#28441]
## 🐛 Bug

Hi there, I am trying to apply backward/grad on the dirichlet.log-prob(x). The input(i.e x) is the output of a call to stick-breaking transform. However, grad output includes nan and -inf for the following x (note that x may not be precise since it comes from execution of my code) 
x = tensor([1.7647e-01, 7.2429e-01, 1.1665e-39, 9.9239e-02]

It seems an overflow for log_prob happens since one of the elements of x is really small. 
## To Reproduce

Here is the code snippet:

```
    transform = dist.StickBreakingTransform()
    x_unconstrained.to(dtype=torch.float64)
    x_unconstrained.requires_grad_(True)

    x_constrained = transform._call(x_unconstrained)
    x_constrained.to(dtype=torch.float64)

    alpha = tensor([2.5, 10.0, 2.0, 3.0], dtype=torch.float64)
    score = dist.Dirichlet(alpha).log_prob(x_constrained)
    score_gradient = grad(score.clone(), x_unconstrained, create_graph=True)[0]
    print("\nscore wrt x", score_gradient)
```
Here is its output:
<img width="619" alt="Screen Shot 2019-10-22 at 12 49 46 PM" src="https://user-images.githubusercontent.com/56326968/67325956-c5c36980-f4ca-11e9-9f4f-46bfe29d27db.png">

## Expected behavior

I expect to get the actual derivative for dirichlet.log_prob.

**Comment by fmassa** [comment]
Can you add a full runnable example that reproduces the issue? `x_unconstrained` is not defined (I suppose it might be equal to `x`, but it's not 100% clear, given that the shapes do not match)

**Comment by torabin** [comment]
Thanks @fmassa for looking into that. I updated the description, let me know if you still need more info.

**Comment by fritzo** [comment]
As a quick workaround, you should be able to `x.clamp_(min=1e-8)` or similar.

**Comment by fehiepsi** [comment]
@torabin I think the issue happens because your `x_unconstrained` takes a so small value `-8000`. Many transforms will give `nan` if we start with "bad" `x_unconstrained` (e.g. ExpTransform). How did you get `x_unconstrained=-8000`? If the initial steps of your inference are unstable, [clip_grad_norm](https://pytorch.org/docs/stable/nn.html#clip-grad-norm) can be helpful.

@neerajprad Increasing threshold might not be helpful in general. The following example shows that even `min=1e-10` is not good (in float32).
```
def _clipped_sigmoid(x):
    finfo = torch.finfo(x.dtype)
    return torch.clamp(torch.sigmoid(x), min=1e-10, max=1. - finfo.eps)

def f(x):
    offset = x.shape[-1] + 1 - x.new_ones(x.shape[-1]).cumsum(-1)
    z = _clipped_sigmoid(x - offset.log())
    z_cumprod = (1 - z).cumprod(-1)
    y = torch.nn.functional.pad(z, (0, 1), value=1) * torch.nn.functional.pad(z_cumprod, (1, 0), value=1)
    return y

x_unconstrained = torch.tensor([1e6, 1e6, 1e6, 1e6, 1e6, -1e6, -1e6, -1e6, -1e6, -1e6], requires_grad=True)
x_constrained = f(x_unconstrained)
score1_gradient = grad(x_constrained.log().sum(), x_unconstrained, create_graph=True)[0]
print("grad", score1_gradient)
```
